python/cudf/cudf/_lib/parquet.pyx

```diff
@@ -767,6 +769,244 @@ cdef class ParquetWriter:
+            self.per_file_user_data = c_result.metadata.per_file_user_data
+
+            # Access the Parquet per_file_user_data to find the index
+
+            self.column_index_type = None
+            self.index_col_names = None
+            self.is_range_index = True
+            for single_file in self.per_file_user_data:
+                json_str = single_file[b'pandas'].decode('utf-8')
+                if json_str != "":
+                    self.pandas_meta = json.loads(json_str)
+                    file_is_range_index, self.index_col, self.column_index_type = \
+                        _parse_metadata(self.pandas_meta)
+                    self.is_range_index &= file_is_range_index
+
+                    if not file_is_range_index and self.index_col is not None \
+                            and self.index_col_names is None:
+                        self.index_col_names = {}
+                        for idx_col in self.index_col:
+                            for c in self.pandas_meta['columns']:
+                                if c['field_name'] == idx_col:
+                                    self.index_col_names[idx_col] = c['name']
+
+        df = cudf.DataFrame._from_data(*data_from_unique_ptr(
+            move(c_result.tbl),
+            column_names=self.names,
+        ))
+
+        self.initialized = True
+        return df
+
+    def read(self):
+        dfs = []
+        while self._has_next():
+            dfs.append(self._read_chunk())
+        df = cudf.concat(dfs)
+        update_struct_field_names(df, self.result_meta.schema_info)
+        if self.pandas_meta is not None:
+            # Book keep each column metadata as the order
+            # of `meta["columns"]` and `column_names` are not
+            # guaranteed to be deterministic and same always.
+            meta_data_per_column = {
+                col_meta['name']: col_meta for col_meta in self.pandas_meta["columns"]
+            }
+
+            # update the decimal precision of each column
+            for col in self.names:
+                if isinstance(df._data[col].dtype, cudf.core.dtypes.DecimalDtype):
+                    df._data[col].dtype.precision = (
+                        meta_data_per_column[col]["metadata"]["precision"]
+                    )
+
+        # Set the index column
+        if self.index_col is not None and len(self.index_col) > 0:
+            if self.is_range_index:
+                if not self.allow_range_index:
+                    return df
+
+                if len(self.per_file_user_data) > 1:
+                    range_index_meta = {
+                        "kind": "range",
+                        "name": None,
+                        "start": 0,
+                        "stop": len(df),
+                        "step": 1
+                    }
+                else:
+                    range_index_meta = self.index_col[0]
+
+                if self.row_groups is not None:
```

Comment:
Is any of the implementation shareable with the non-chunked parquet reader?

Reply:
Made the implementations share almost all of the code.